File: backend/app/services/auth_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.models.user import User
from app.core.security import verify_password, get_password_hash, create_access_token, create_refresh_token
from app.schemas.user import UserCreate
from fastapi import HTTPException, status
import uuid


async def get_user_by_email(db: AsyncSession, email: str):
    result = await db.execute(select(User).where(User.email == email))
    return result.scalar_one_or_none()


async def get_user_by_id(db: AsyncSession, user_id: str):
    result = await db.execute(select(User).where(User.id == uuid.UUID(user_id)))
    return result.scalar_one_or_none()


async def create_user(db: AsyncSession, data: UserCreate):
    existing = await get_user_by_email(db, data.email)
    if existing:
        raise HTTPException(status_code=400, detail="Email already registered")
    user = User(
        full_name=data.full_name,
        email=data.email,
        password_hash=get_password_hash(data.password),
        role=data.role,
        student_id=data.student_id,
        phone_number=data.phone_number,
    )
    db.add(user)
    await db.commit()
    await db.refresh(user)
    return user


async def authenticate_user(db: AsyncSession, email: str, password: str):
    user = await get_user_by_email(db, email)
    if not user or not verify_password(password, user.password_hash):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect email or password",
        )
    if not user.is_active:
        raise HTTPException(status_code=400, detail="Account is inactive")
    return user


async def generate_tokens(user: User):
    access_token = create_access_token({"sub": str(user.id), "role": user.role})
    refresh_token = create_refresh_token({"sub": str(user.id)})
    return access_token, refresh_token

File: backend-php/src/Services/AuthService.php
<?php
declare(strict_types=1);

namespace App\Services;

use App\Core\Database;
use App\Core\Security;
use PDO;

class AuthService
{
    public static function getUserByEmail(string $email): ?array
    {
        $pdo  = Database::getConnection();
        $stmt = $pdo->prepare('SELECT * FROM users WHERE email = ?');
        $stmt->execute([$email]);
        return $stmt->fetch() ?: null;
    }

    public static function getUserById(string $userId): ?array
    {
        $pdo  = Database::getConnection();
        $stmt = $pdo->prepare('SELECT * FROM users WHERE id = ?');
        $stmt->execute([$userId]);
        return $stmt->fetch() ?: null;
    }

    public static function createUser(array $data): array
    {
        $existing = self::getUserByEmail($data['email']);
        if ($existing) {
            throw new \RuntimeException('Email already registered', 400);
        }

        $pdo  = Database::getConnection();
        $stmt = $pdo->prepare('
            INSERT INTO users (full_name, email, password_hash, role, student_id, phone_number)
            VALUES (?, ?, ?, ?, ?, ?)
            RETURNING *
        ');
        $stmt->execute([
            $data['full_name'],
            $data['email'],
            Security::hashPassword($data['password']),
            $data['role'] ?? 'student',
            $data['student_id'] ?? null,
            $data['phone_number'] ?? null,
        ]);
        return $stmt->fetch();
    }

    public static function authenticateUser(string $email, string $password): array
    {
        $user = self::getUserByEmail($email);
        if (!$user || !Security::verifyPassword($password, $user['password_hash'])) {
            throw new \RuntimeException('Incorrect email or password', 401);
        }
        if (!$user['is_active']) {
            throw new \RuntimeException('Account is inactive', 400);
        }
        return $user;
    }

    public static function generateTokens(array $user): array
    {
        $accessToken  = Security::createAccessToken(['sub' => $user['id'], 'role' => $user['role']]);
        $refreshToken = Security::createRefreshToken(['sub' => $user['id']]);
        return [$accessToken, $refreshToken];
    }
}
